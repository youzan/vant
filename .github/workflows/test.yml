name: CI

on:
  push:
    branches:
      - '**'

  pull_request:
    branches:
      - main

  workflow_dispatch:

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4

      - name: Install pnpm
        run: |
          npm install -g corepack@latest --force 
          corepack enable

      - uses: actions/setup-node@v4
        with:
          node-version: '22.14.0'
          cache: 'pnpm'

      - name: Install dependencies
        run: pnpm install

      - name: Run linter
        run: npm run lint

  test:
    runs-on: macos-14
    steps:
      - uses: actions/checkout@v4

      - name: Install pnpm
        run: |
          npm install -g corepack@latest --force 
          corepack enable

      - uses: actions/setup-node@v4
        with:
          node-version: '22.14.0'
          cache: 'pnpm'

      - name: Install dependencies
        run: pnpm install

      - name: Run test cases
        run: pnpm run test:coverage

      - name: Upload coverage to Codecov
        uses: codecov/codecov-action@v5
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          directory: ./packages/vant/test/coverage

  build:
    runs-on: macos-14
    steps:
      - uses: actions/checkout@v4

      - name: Install pnpm
        run: |
          npm install -g corepack@latest --force 
          corepack enable

      - uses: actions/setup-node@v4
        with:
          node-version: '22.14.0'
          cache: 'pnpm'

      - name: Install dependencies
        run: pnpm install

      - name: Build
        run: npm run build
