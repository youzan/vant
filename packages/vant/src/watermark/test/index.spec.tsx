// @ts-nocheck
import { Watermark } from '..';
import { mount } from '../../../test';
import { KeepAlive } from 'vue';

describe('watermark', () => {
  beforeEach(() => {
    const createElement = document.createElement.bind(document);
    document.createElement = (tagName: string) => {
      if (tagName === 'canvas') {
        return {
          ...createElement(tagName),
          getContext: () => {
            () => {};
          },
          toDataURL: () => 'base64Url',
        };
      }
      return createElement(tagName);
    };
    global.URL.createObjectURL = vi.fn(() => 'blob:test');
    global.Image = class {
      crossOrigin = 'anonymous';

      referrerPolicy = 'no-referrer';

      naturalWidth = 800;

      naturalHeight = 550;

      onload: () => void = () => {};

      // just mock to trigge onload
      _src = '';

      get src() {
        return this._src;
      }

      set src(val) {
        this._src = val;
        this.onload();
      }
    };
  });

  test('should render content', () => {
    const wrapper = mount(Watermark, {
      props: {
        content: 'Vant',
        textColor: 'red',
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should render image', () => {
    const wrapper = mount(Watermark, {
      props: {
        content: 'Vant',
        image:
          'https://fastly.jsdelivr.net/npm/@vant/assets/vant-watermark.png',
        opacity: 0.5,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should render html', () => {
    const wrapper = mount(Watermark, {
      slots: {
        content: () => 'vant watermark test',
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test('test width, height, rotate, zIndex', () => {
    const wrapper = mount(Watermark, {
      props: {
        width: 20,
        height: 20,
        rotate: 20,
        zIndex: 200,
      },
      slots: {
        content: () => 'vant watermark test',
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test('test false value fullPage', () => {
    const wrapper = mount(Watermark, {
      props: {
        fullPage: false,
      },
      slots: {
        content: () => 'vant watermark test',
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should render KeepAlive with Watermark correctly', async () => {
    const wrapper = mount({
      render() {
        return <KeepAlive>{this.render ? <Watermark /> : null}</KeepAlive>;
      },
      data() {
        return {
          render: false,
        };
      },
    });

    await wrapper.setData({ render: true });
    expect(wrapper.find('.van-watermark').style.backgroundImage).toBe(
      'url(blob:test)',
    );
  });
});
