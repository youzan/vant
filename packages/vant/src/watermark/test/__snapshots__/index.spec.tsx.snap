// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`watermark > should render content 1`] = `
<div
  class="van-watermark van-watermark--full"
  style="background-image: url();"
>
  <div class="van-watermark__wrapper">
    <svg
      viewbox="0 0 100 100"
      width="100"
      height="100"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
      style="padding: 0px 0px 0px 0px;"
    >
      <foreignObject
        x="0"
        y="0"
        width="100"
        height="100"
      >
        <div
          xmlns="http://www.w3.org/1999/xhtml"
          style="transform-origin: center; transform: rotate(-22deg);"
        >
          <span style="color: red;">
            Vant
          </span>
        </div>
      </foreignObject>
    </svg>
  </div>
</div>
`;

exports[`watermark > should render html 1`] = `
<div
  class="van-watermark van-watermark--full"
  style="background-image: url();"
>
  <div class="van-watermark__wrapper">
    <svg
      viewbox="0 0 100 100"
      width="100"
      height="100"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
      style="padding: 0px 0px 0px 0px;"
    >
      <foreignObject
        x="0"
        y="0"
        width="100"
        height="100"
      >
        <div
          xmlns="http://www.w3.org/1999/xhtml"
          style="transform-origin: center; transform: rotate(-22deg);"
        >
          vant watermark test
        </div>
      </foreignObject>
    </svg>
  </div>
</div>
`;

exports[`watermark > should render image 1`] = `
<div
  class="van-watermark van-watermark--full"
  style="background-image: url();"
>
  <div class="van-watermark__wrapper">
    <svg
      viewbox="0 0 100 100"
      width="100"
      height="100"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
      style="padding: 0px 0px 0px 0px; opacity: 0.5;"
    >
      <image
        href="base64Url"
        xlink:href="base64Url"
        x="0"
        y="0"
        width="100"
        height="100"
        style="transform-origin: center; transform: rotate(-22deg);"
      >
      </image>
    </svg>
  </div>
</div>
`;

exports[`watermark > test false value fullPage 1`] = `
<div
  class="van-watermark"
  style="background-image: url();"
>
  <div class="van-watermark__wrapper">
    <svg
      viewbox="0 0 100 100"
      width="100"
      height="100"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
      style="padding: 0px 0px 0px 0px;"
    >
      <foreignObject
        x="0"
        y="0"
        width="100"
        height="100"
      >
        <div
          xmlns="http://www.w3.org/1999/xhtml"
          style="transform-origin: center; transform: rotate(-22deg);"
        >
          vant watermark test
        </div>
      </foreignObject>
    </svg>
  </div>
</div>
`;

exports[`watermark > test width, height, rotate, zIndex 1`] = `
<div
  class="van-watermark van-watermark--full"
  style="background-image: url(); z-index: 200;"
>
  <div class="van-watermark__wrapper">
    <svg
      viewbox="0 0 20 20"
      width="20"
      height="20"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
      style="padding: 0px 0px 0px 0px;"
    >
      <foreignObject
        x="0"
        y="0"
        width="20"
        height="20"
      >
        <div
          xmlns="http://www.w3.org/1999/xhtml"
          style="transform-origin: center; transform: rotate(20deg);"
        >
          vant watermark test
        </div>
      </foreignObject>
    </svg>
  </div>
</div>
`;
